File: Gfg_list/Microsoft_30_Days/Day5/1-permutation_of_string.cpp



// https://practice.geeksforgeeks.org/batch/microsdaysoft-29/track/microsoft-29days-day4/problem/top-k-frequent-elements-in-array







#include <bits/stdc++.h>
#include<ext/pb_ds/tree_policy.hpp>
#include<ext/pb_ds/assoc_container.hpp>


using namespace std;
using namespace __gnu_pbds ;
typedef  long long int ll;
#define DEBUG_var 1
#define fastIO ios_base::sync_with_stdio(false);cin.tie(0);cout.tie(0);


void rec_uitl(int ind , string S , set<string> &st1)
{
	if(S.size()==ind)
		st1.insert(S);

	for(int it = ind ; it < S.size() ; it++)
	{
		swap(S[ind] , S[it]);
		rec_uitl(ind+1 , S , st1);
		swap(S[ind] , S[it]);

	}
}
vector<string>find_permutation(string S)
{
	// Code here there
	set<string> ans2  ;

	rec_uitl(0 , S , ans2);
	vector<string> ans1(ans2.begin(),ans2.end())  ;

	return ans1 ;
}

File: Gfg_list/Microsoft_30_Days/Day5/1_permutation_of_string.py
# https://practice.geeksforgeeks.org/batch/microsdaysoft-29/track/microsoft-29days-day5/problem/permutations-of-a-given-string2041

def permute(S , l1 ,r1,ans1):
    if(l1==r1):
        ans1.append(''.join(S))
    else :
        for i in range(l1,r1):
            S[l1] , S[i] = S[i],S[l1]
            permute(S,l1+1 , r1 , ans1)
            S[l1] , S[i] = S[i],S[l1]



class Solution:
    def find_permutation(self, S):
        # Code here
        ans1 = []
        n = len(S)
        S = list(S)
        permute(S , 0 , n,ans1)
        ans1 =set(ans1)
        ans1 = list(ans1)
        ans1.sort()
        return ans1 
        






